README.md
# Slp parser for dart

This package is for parsing [Simple Ledger Protocol (SLP)](https://github.com/simpleledger/slp-specifications) metadata. TokenType1 and NFT1 tokens are supported.


## Usage

```dart
  var scriptPubKey = '6a04534c500001010747454e455349530ce3838ae382abe383a2e383884c004c004c0001094c00080000000000000064';
  var slpMsg = parseSLP(hex.decode(scriptPubKey));
  print('Parsed results (raw form):\n ${slpMsg.toMap(raw: true)}');
  print('Parsed results (readable form):\n ${slpMsg.toMap()}');
```

## Test

`$ dart test/parser.spec.dart`



CHANGELOG.md
## 0.1.0
### Added
    * Null safety update
    * Travis config added

## 0.0.3
### Added
    * Implemented classes and toMap methods
    * Example file
### Modified
    * Fixed linting issues
    * Deleted .ts file used during porting process

## 0.0.2
### Modified
    * Updated pubspec

## 0.0.1
### Added
    * Initial version with 101 unit tests passing (no fuzzer tests yet)


test/parser.spec.dart
import 'package:test/test.dart';
import 'package:convert/convert.dart';

import 'package:slp_parser/slp_parser.dart';
import './script_tests.dart';

void main() {
  testdata.forEach((testcase) {
    test(testcase["msg"], () {
      if (testcase["code"] == null) {
        final ParseResult slpMsg = parseSLP(hex.decode(testcase["script"] as String));
        final Map<String, Object> exp = testcase["parsed"] as Map<String, Object>;
        final Map<String, Object> data = exp["data"] as Map<String, Object>;
        if (slpMsg.transactionType == "GENESIS") {
          data["qty"] = BigInt.parse(data["qty"] as String);
        } else if (slpMsg.transactionType == "MINT") {
          data["tokenId"] = hex.decode(data["tokenId"] as String);
          data["qty"] = BigInt.parse(data["qty"] as String);
        } else if (slpMsg.transactionType == "SEND") {
          data["tokenId"] = hex.decode(data["tokenId"] as String);
          var amounts = (data["amounts"] as List<String>).map((a) { return BigInt.parse(a);}).toList() as List<BigInt>;
          data["amounts"] = amounts;
        }
        exp["data"] = data;
        final slpMsgMap = slpMsg.toMap(raw: true);
        expect(slpMsgMap, exp);
      } else {
        var throws = true;
        try {
          parseSLP(hex.decode(testcase["script"] as String));
          throws = false;
        } catch(_) {
          expect(throws, true);
        }
      }
    });
  });
}


fuzz/server.dart
import 'dart:convert';
import 'dart:io';
import 'package:slp_parser/slp_parser.dart';
import 'package:convert/convert.dart';

Future main() async {
  var server = await HttpServer.bind(
    InternetAddress.loopbackIPv4,
    8080,
  );
  print('Listening on localhost: ${server.port}');

  await for (HttpRequest request in server) {
    print(request.uri.toString());

    var y = null;
    try {
      var url = request.uri.toString();
      var bin = hex.decode(url[0] == '/' ? url.substring(1) : url);
      y = parseSLP(bin);
      y.versionType = y.tokenType;

      if (y.transactionType == "GENESIS") {
          y.symbol         = hex.encode(y.data.ticker).toUpperCase();
          y.name           = hex.encode(y.data.name).toUpperCase();
          y.documentUri    = hex.encode(y.data.documentUri).toUpperCase();
          if (y.data.documentHash == null) {
              y.documentSha256 = "";
          }
          y.documentSha256 = hex.encode(y.data.documentHash).toUpperCase();
      }

      if (y.data.hasOwnProperty("decimals")) {
          y.decimals = y.data.decimals;
      }

      if (y.data.hasOwnProperty("qty")) {
          y.genesisOrMintQuantity = (y.data.qty).toString();
      }

      if (y.data.hasOwnProperty("tokenid")) {
          y.tokenIdHex = hex.encode(y.data.tokenid);
      }

      if (y.data.hasOwnProperty("mintBatonVout")) {
          y.batonVout = y.data.mintBatonVout;
      }

      if (y.data.hasOwnProperty("amounts")) {
          y.sendOutputs = y.data.amounts.map((v) => v.toString());
      }
    } catch(e) {
      var buf = utf8.encode("{'success': false, 'error:': ${e}}");
      request.headers.add('Content-Length', buf.length);
      request.response.write(buf);
    }

    await request.response.close();
  }
}


lib/src/index.ts

export interface GenesisParseResult {
  ticker: Buffer,
  name: Buffer,
  documentUri: Buffer,
  documentHash: Buffer,
  decimals: number,
  mintBatonVout: number,
  qty: BN
};

export interface MintParseResult {
  tokenId: Buffer,
  mintBatonVout: number,
  qty: BN
};

export interface SendParseResult {
  tokenId: Buffer,
  amounts: BN[]
};

interface ImpossibleParseResult {
};

export interface ParseResult {
  tokenType: number,
  transactionType: string,
  data: GenesisParseResult | MintParseResult | SendParseResult | ImpossibleParseResult
};

export const parseSLP = (scriptpubkey: Buffer | string): ParseResult => {
  if (typeof scriptpubkey === "string") {
    scriptpubkey = Buffer.from(scriptpubkey, 'hex');
  }
  let it: number = 0; // position in itObj
  let itObj: Buffer = scriptpubkey; // object it refers to


  const OP_0: number         = 0x00;
  const OP_16: number        = 0x60;
  const OP_RETURN: number    = 0x6a;
  const OP_PUSHDATA1: number = 0x4c;
  const OP_PUSHDATA2: number = 0x4d;
  const OP_PUSHDATA4: number = 0x4e;

  const PARSE_CHECK = (v: boolean, str: string): void => {
    if (v) {
      throw Error(str);
    }
  };

  const extractU8 = (): BN => {
    const r: number = itObj.readUInt8(it);
    it += 1;
    return new BN(r);
  };

  const extractU16 = (): BN => {
    const r: number= itObj.readUInt16LE(it);
    it += 2;
	return new BN(r);
  };

  const extractU32 = (): BN => {
    const r: number = itObj.readUInt32LE(it);
	it += 4;
	return new BN(r);
  };

  const extractU64 = (): BN => {
    const r1: number = itObj.readUInt32LE(it);
	  it += 4;

    const r2: number = itObj.readUInt32LE(it);
	  it += 4;

	  return new BN(r2).multipliedBy(2**32).plus(r1);
  };

  PARSE_CHECK(itObj.length === 0, "scriptpubkey cannot be empty");
  PARSE_CHECK(itObj[it] !== OP_RETURN, "scriptpubkey not op_return");
  PARSE_CHECK(itObj.length < 10, "scriptpubkey too small"); // TODO what is correct minimum size?
  ++it;

  const extractPushdata = (): number => {
    if (it === itObj.length) {
      return -1;
    }

    const cnt: number = extractU8().toNumber();
    if (cnt > OP_0 && cnt < OP_PUSHDATA1) {
      if (it+cnt > itObj.length) {
        --it; return -1;
      }

      return cnt;
    }
    else if (cnt === OP_PUSHDATA1) {
      if (it+1 >= itObj.length) {
        --it; return -1;
      }
      return extractU8().toNumber();
    }
    else if (cnt === OP_PUSHDATA2) {
      if (it+2 >= itObj.length) {
        --it; return -1;
      }
      return extractU16().toNumber();
    }
    else if (cnt === OP_PUSHDATA4) {
      if (it+4 >= itObj.length) {
        --it; return -1;
      }
      return extractU32().toNumber();
    }

    // other opcodes not allowed
    --it; return -1;
  };

  const bufferToBN = (): BN => {
    if (itObj.length === 1) return extractU8();
    if (itObj.length === 2) return extractU16();
    if (itObj.length === 4) return extractU32();
    if (itObj.length === 8) return extractU64();
	throw new Error('extraction of number from buffer failed');
  };

  const checkValidTokenId = (tokenId: Buffer): boolean => tokenId.length === 32;

  const chunks: Buffer[] = [];
  for (let len = extractPushdata(); len >= 0; len = extractPushdata()) {
    const buf: Buffer = itObj.slice(it, it+len);
    PARSE_CHECK(it + len > itObj.length, "pushdata data extraction failed");

    it += len;
    chunks.push(buf);

    if (chunks.length === 1) {
      const lokadIdStr = chunks[0];
      PARSE_CHECK(lokadIdStr.length !== 4, "lokad id wrong size");
      PARSE_CHECK(
          lokadIdStr[0] !== 'S'.charCodeAt(0)
       || lokadIdStr[1] !== 'L'.charCodeAt(0)
       || lokadIdStr[2] !== 'P'.charCodeAt(0)
       || lokadIdStr[3] !== 0x00, "SLP not in first chunk");
    }
  }

  PARSE_CHECK(it !== itObj.length, "trailing data");
  PARSE_CHECK(chunks.length === 0, "chunks empty");

  let cit = 0;
  const CHECK_NEXT = (): void => {
    ++cit;
    PARSE_CHECK(cit === chunks.length, "parsing ended early");
    it = 0;
    itObj = chunks[cit];
  };
  CHECK_NEXT(); // for quick exit check done above

  const tokenTypeBuf = itObj.reverse();
  PARSE_CHECK(tokenTypeBuf.length !== 1 && tokenTypeBuf.length !== 2,
      "token_type string length must be 1 or 2");

  const tokenType: number = bufferToBN().toNumber();
  PARSE_CHECK(! [0x01, 0x41, 0x81].includes(tokenType),
              "token_type not token-type1, nft1-group, or nft1-child");

  CHECK_NEXT();

  const transactionType = itObj.toString();
  if (transactionType === 'GENESIS') {
    PARSE_CHECK(chunks.length !== 10, "wrong number of chunks");
    CHECK_NEXT();

    const ticker = itObj;
    CHECK_NEXT();

    const name = itObj;
    CHECK_NEXT();

    const documentUri = itObj;
    CHECK_NEXT();

    const documentHash = itObj;
    PARSE_CHECK(! (documentHash.length === 0 || documentHash.length === 32),
      "document_hash must be size 0 or 32");
    CHECK_NEXT();

    const decimalsBuf = itObj;
    PARSE_CHECK(decimalsBuf.length !== 1, "decimals string length must be 1");

    const decimals = bufferToBN().toNumber();
    PARSE_CHECK(decimals > 9, "decimals bigger than 9");
    CHECK_NEXT();

    const mintBatonVoutBuf = itObj;
    let mintBatonVout = 0;
    PARSE_CHECK(mintBatonVoutBuf.length >= 2, "mint_baton_vout string length must be 0 or 1");
    if (mintBatonVoutBuf.length > 0) {
      mintBatonVout = bufferToBN().toNumber();
      PARSE_CHECK(mintBatonVout < 2, "mint_baton_vout must be at least 2");
    }
    CHECK_NEXT();

    const qtyBuf= itObj.reverse();
    PARSE_CHECK (qtyBuf.length !== 8, "initial_qty must be provided as an 8-byte buffer");
    const qty = bufferToBN();

    if (tokenType === 0x41) {
      PARSE_CHECK(decimals !== 0, "NFT1 child token must have divisibility set to 0 decimal places");
      PARSE_CHECK(mintBatonVout !== 0, "NFT1 child token must not have a minting baton");
      PARSE_CHECK(! qty.isEqualTo(1), "NFT1 child token must have quantity of 1");
    }

	const actionData: GenesisParseResult = {
	  ticker,
	  name,
      documentUri,
      documentHash,
      decimals,
      mintBatonVout,
      qty
	};

    return {
      tokenType,
      transactionType,
      data: actionData
    };
  } else if (transactionType === "MINT") {
    PARSE_CHECK(tokenType === 0x41, "NFT1 Child cannot have MINT transaction type.");

    PARSE_CHECK(chunks.length !== 6, "wrong number of chunks");
    CHECK_NEXT();

    const tokenId: Buffer = itObj;
    PARSE_CHECK(! checkValidTokenId(tokenId), "tokenId invalid size");
    CHECK_NEXT();

    const mintBatonVoutBuf = itObj;
    let mintBatonVout = 0;
    PARSE_CHECK(mintBatonVoutBuf.length >= 2, "mint_baton_vout string length must be 0 or 1");
    if (mintBatonVoutBuf.length > 0) {
      mintBatonVout = bufferToBN().toNumber();
      PARSE_CHECK(mintBatonVout < 2, "mint_baton_vout must be at least 2");
    }
    CHECK_NEXT();

    const additionalQtyBuf = itObj.reverse();
    PARSE_CHECK (additionalQtyBuf.length !== 8, "additional_qty must be provided as an 8-byte buffer");
    const qty = bufferToBN();

    const actionData: MintParseResult = {
      tokenId,
      mintBatonVout,
      qty
    };

    return {
      tokenType,
      transactionType,
      data: actionData
    };
  } else if (transactionType === "SEND") {
    PARSE_CHECK(chunks.length < 4, "wrong number of chunks");
    CHECK_NEXT();

    const tokenId: Buffer = itObj;
    PARSE_CHECK(! checkValidTokenId(tokenId), "tokenId invalid size");
    CHECK_NEXT();

    const amounts: BN[] = [];
    while (cit !== chunks.length) {
      const amountBuf = itObj.reverse();
      PARSE_CHECK(amountBuf.length !== 8, "amount string size not 8 bytes");

      const value: BN = bufferToBN();
      amounts.push(value);

      ++cit;
      itObj = chunks[cit];
      it = 0;
    }

    PARSE_CHECK(amounts.length === 0, "token_amounts size is 0");
    PARSE_CHECK(amounts.length > 19, "token_amounts size is greater than 19");

    const actionData: SendParseResult = {
      tokenId,
      amounts
    };

    return {
      tokenType,
      transactionType,
      data: actionData
    };
  } else {
	PARSE_CHECK(true, "unknown action type");
  }

  // unreachable code
  return {
    tokenType,
    transactionType,
    data: {}
  };
};

export {
  BN
};

example/main.dart
import 'package:slp_parser/slp_parser.dart';
import 'package:convert/convert.dart';

main() {
  var scriptPubKey = '6a04534c500001010747454e455349530ce3838ae382abe383a2e383884c004c004c0001094c00080000000000000064';
  var slpMsg = parseSLP(hex.decode(scriptPubKey));
  print('Parsed results (raw form):\n ${slpMsg.toMap(raw: true)}');
  print('Parsed results (readable form):\n ${slpMsg.toMap()}');

  scriptPubKey = '6a04534c500001010453454e44208888888888888888888888888888888888888888888888888888888888888888080000000000000001080000000000000001080000000000000001080000000000000001080000000000000001080000000000000001080000000000000001080000000000000001080000000000000001080000000000000001080000000000000001080000000000000001080000000000000001080000000000000001080000000000000001080000000000000001080000000000000001080000000000000001080000000000000001';
  slpMsg = parseSLP(hex.decode(scriptPubKey));
  print('Parsed results (raw form):\n ${slpMsg.toMap(raw: true)}');
  print('Parsed results (readable form):\n ${slpMsg.toMap()}');
}


lib/src/index.dart
import 'dart:typed_data';
import 'dart:convert';
import 'package:convert/convert.dart';

class GenesisParseResult {
  List<int> tickerBuf, nameBuf, documentUriBuf, documentHashBuf;
  int decimals;
  int mintBatonVout;
  BigInt qty;
  String get ticker { return utf8.decode(this.tickerBuf); }
  String get name { return utf8.decode(this.nameBuf); }
  String get documentUri { return utf8.decode(this.documentUriBuf); }
  String get documentHash { return hex.encode(this.documentHashBuf); }
  GenesisParseResult({
    required this.tickerBuf,
    required this.nameBuf,
    required this.documentUriBuf,
    required this.documentHashBuf,
    required this.decimals,
    required this.mintBatonVout,
    required this.qty
  });
  Map<String, Object> toMap({raw=false}) {
    return {
      "ticker": raw ? this.tickerBuf : this.ticker,
      "name": raw ? this.nameBuf : this.name,
      "documentUri": raw ? this.documentUriBuf : this.documentUri,
      "documentHash": raw ? this.documentHashBuf : this.documentHash,
      "decimals": this.decimals,
      "mintBatonVout": this.mintBatonVout,
      "qty": this.qty
    };
  }
}

class MintParseResult {
  List<int> tokenIdBuf;
  int mintBatonVout;
  BigInt qty;
  String get tokenId { return hex.encode(this.tokenIdBuf); }
  MintParseResult({
    required this.tokenIdBuf,
    required this.mintBatonVout,
    required this.qty
  });
  Map<String, Object> toMap({raw=false}) {
    return {
      "tokenId": raw ? this.tokenIdBuf : this.tokenId,
      "mintBatonVout": mintBatonVout,
      "qty": this.qty
    };
  }
} 

class SendParseResult {
  List<int> tokenIdBuf;
  List<BigInt> amounts;
  SendParseResult({
    required this.tokenIdBuf,
    required this.amounts
  });
  String get tokenId { return hex.encode(this.tokenIdBuf); }
  Map<String, Object> toMap({raw=false}) {
    return {
      "tokenId": raw ? this.tokenIdBuf : this.tokenId,
      "amounts": this.amounts
    };
  }
}

class ImpossibleParseResult {
}

class ParseResult {
  int tokenType;
  String transactionType;
  dynamic data; /* GenesisParseResult | MintParseResult | SendParseResult | ImpossibleParseResult */
  ParseResult({
    required this.tokenType,
    required this.transactionType,
    this.data
  });
  Map<String, Object> toMap({raw: false}) {
    return {
      "tokenType": tokenType,
      "transactionType": transactionType,
      "data": data.toMap(raw: raw)
    };
  }
}

ParseResult parseSLP(List<int> scriptPubKey) {
  int it = 0;
  var itObj = Uint8List(scriptPubKey.length);
  itObj.setAll(0, scriptPubKey);

  const int OP_0 = 0x00;
  const int OP_RETURN = 0x6a;
  const int OP_PUSHDATA1 = 0x4c;
  const int OP_PUSHDATA2 = 0x4d;
  const int OP_PUSHDATA4 = 0x4e;

  final PARSE_CHECK = /* void */ (bool v , String str ) {
    if (v) {
      throw(str);
    }
  };

  final extractU8 = /* BN */ () {
    var r = itObj.buffer.asByteData().getUint8(it);
    it += 1;
    return r;
  };

  final extractU16 = /* BN */ (Endian endian) { 
    var r = itObj.buffer.asByteData().getUint16(it, endian);
    it += 2;
    return r;
  };

  final extractU32 = /* BN */ (Endian endian) {
    var r = itObj.buffer.asByteData().getUint32(it, endian);
    it += 4;
    return r;
  };

  final extractU64 = /* BN */ (Endian endian) {
    var r1 = itObj.buffer.asByteData().getUint32(it, endian);
    it += 4;
    var r2 = itObj.buffer.asByteData().getUint32(it, endian);
    it += 4;
    var b1 = BigInt.from(r1);
    var b2 = BigInt.from(r2);
    if (endian == Endian.big) {
      return (b1 * BigInt.from(2).pow(32)) + b2;
    } else {
      return (b2 * BigInt.from(2).pow(32)) + b1;
    }
  };

  PARSE_CHECK(itObj.lengthInBytes == 0, "scriptpubkey cannot be empty");
  PARSE_CHECK(itObj.buffer.asByteData().getUint8(it) != OP_RETURN, "scriptpubkey not op_return" );
  PARSE_CHECK(itObj.lengthInBytes < 10 ,"scriptpubkey too small" );
  ++it;

  final extractPushdata = /* num */ () {
    if (it == itObj.length) { return -1;}
    final int cnt = extractU8();
    if (cnt > OP_0 && cnt < OP_PUSHDATA1) { 
      if (it + cnt > itObj.length) { --it; return -1;} 
      return cnt;
    } else if (cnt == OP_PUSHDATA1) {
      if (it + 1 >= itObj.length) { --it; return -1;} 
      return extractU8();
    } else if (cnt == OP_PUSHDATA2) {
      if (it + 2 >= itObj.length ) { --it; return -1;} 
      return extractU16(Endian.little);
    } else if (cnt == OP_PUSHDATA4) { 
      if (it + 4 >= itObj.length ) { --it; return -1;}
      return extractU32(Endian.little);
    }

    // other opcodes not allowed
    --it;
    return -1;
  };

  final bufferToBN = /* BN */ () {
    if (itObj.length == 1) {
      return BigInt.from(extractU8());
    }
    if (itObj.length == 2) {
      return BigInt.from(extractU16(Endian.big));
    }
    if (itObj.length == 4) {
      return BigInt.from(extractU32(Endian.big));
    }
    if (itObj.length == 8) {
      return extractU64(Endian.big);
    }
    throw("extraction of number from buffer failed");
  };

  final checkValidTokenId = (Uint8List tokenId) => tokenId.length == 32;
  final List<Uint8List> chunks = [];
  for (var len = extractPushdata();len >= 0;len = extractPushdata()) {
    final buf = Uint8List(len);
    buf.setAll(0, itObj.getRange(it, it + len));
    PARSE_CHECK(it + len > itObj.length, "pushdata data extraction failed" );
    it += len;
    chunks.add(buf);
    if (chunks.length == 1) {
      final lokadIdStr = chunks[0];
      PARSE_CHECK(lokadIdStr.length != 4, "lokad id wrong size");
      PARSE_CHECK(
          lokadIdStr.buffer.asByteData().getUint8(0) != "S".codeUnitAt(0)
        || lokadIdStr.buffer.asByteData().getUint8(1) != "L".codeUnitAt(0)
        || lokadIdStr.buffer.asByteData().getUint8(2) != "P".codeUnitAt(0)
        || lokadIdStr.buffer.asByteData().getUint8(3) != 0x00, "SLP not in first chunk");  
    }
  }

  PARSE_CHECK(it != itObj.length, "trailing data");
  PARSE_CHECK(chunks.isEmpty, "chunks empty");
  
  var cit = 0;
  final CHECK_NEXT = /* void */ () {
    ++cit;
    PARSE_CHECK(cit == chunks.length, "parsing ended early");
    it = 0;
    itObj = chunks[cit];
  };
  CHECK_NEXT();

  final tokenTypeBuf = itObj;
  PARSE_CHECK(tokenTypeBuf.length != 1 && tokenTypeBuf.length != 2,
    "token_type string length must be 1 or 2" );
  int tokenType = 0;
  if (tokenTypeBuf.length == 1) {
    tokenType = extractU8();
  } else if (tokenTypeBuf.length == 2) {
    tokenType = extractU16(Endian.big);
  }
  PARSE_CHECK(![ 0x01 , 0x41 , 0x81 ].contains(tokenType),
    "token_type not token-type1, nft1-group, or nft1-child" );
  CHECK_NEXT();

  final transactionType = utf8.decode(itObj);
  if (transactionType == "GENESIS") { 
    PARSE_CHECK(chunks.length != 10, "wrong number of chunks" );
    CHECK_NEXT();

    final ticker = itObj;
    CHECK_NEXT();

    final name = itObj;
    CHECK_NEXT();

    final documentUri = itObj;
    CHECK_NEXT();

    final documentHash = itObj;
    PARSE_CHECK(documentHash.length != 0 && documentHash.length != 32 , "document_hash must be size 0 or 32" );
    CHECK_NEXT();

    final decimalsBuf = itObj;
    PARSE_CHECK(decimalsBuf.length != 1, "decimals string length must be 1" );

    final decimals = bufferToBN().toInt();
    PARSE_CHECK(decimals > 9,
      "decimals bigger than 9" );
    CHECK_NEXT();

    final mintBatonVoutBuf = itObj;
    int mintBatonVout = 0;
    PARSE_CHECK(mintBatonVoutBuf.length >= 2 , "mint_baton_vout string length must be 0 or 1" );
    if (mintBatonVoutBuf.length > 0 ) {
      mintBatonVout = bufferToBN().toInt();
      PARSE_CHECK(mintBatonVout < 2, "mint_baton_vout must be at least 2" );
    }
    CHECK_NEXT();

    final qtyBuf = itObj;//.reversed;
    PARSE_CHECK(qtyBuf.length != 8 ,
      "initial_qty must be provided as an 8-byte buffer" );
    final qty = bufferToBN();

    if (tokenType == 0x41) {
      PARSE_CHECK(decimals != 0, "NFT1 child token must have divisibility set to 0 decimal places" );
      PARSE_CHECK(mintBatonVout != 0, "NFT1 child token must not have a minting baton" );
      PARSE_CHECK(qty.compareTo(BigInt.from(1)) != 0, "NFT1 child token must have quantity of 1" );
    }
    final GenesisParseResult actionData = new GenesisParseResult(
      tickerBuf: ticker,
      nameBuf: name,
      documentUriBuf: documentUri,
      documentHashBuf: documentHash,
      decimals: decimals,
      mintBatonVout: mintBatonVout,
      qty: qty
    );
    return new ParseResult(
      tokenType: tokenType,
      transactionType: transactionType,
      data: actionData
    );
  } else if (transactionType == "MINT") {
    PARSE_CHECK(tokenType == 0x41, "NFT1 Child cannot have MINT transaction type." );

    PARSE_CHECK(chunks.length != 6, "wrong number of chunks");
    CHECK_NEXT();

    final tokenId = itObj;
    PARSE_CHECK(! checkValidTokenId(tokenId), "tokenId invalid size");
    CHECK_NEXT();

    final mintBatonVoutBuf = itObj;
    var mintBatonVout = 0;
    PARSE_CHECK(mintBatonVoutBuf.length >= 2 , "mint_baton_vout string length must be 0 or 1" );
    if (mintBatonVoutBuf.length > 0 ) {
      if (mintBatonVoutBuf.length == 1) {
        mintBatonVout = extractU8();
      } else if (mintBatonVoutBuf.length == 2) {
        mintBatonVout = extractU16(Endian.big);
      }
      PARSE_CHECK(mintBatonVout < 2 , "mint_baton_vout must be at least 2");
    }
    CHECK_NEXT();

    final additionalQtyBuf = itObj;//.reversed;
    PARSE_CHECK(additionalQtyBuf.length != 8, "additional_qty must be provided as an 8-byte buffer" );
    final qty = bufferToBN();

    final actionData = new MintParseResult(
      tokenIdBuf: tokenId,
      mintBatonVout: mintBatonVout,
      qty: qty
    );

    return new ParseResult( 
      tokenType: tokenType,
      transactionType: transactionType,
      data: actionData
    );
  } else if (transactionType == "SEND") {
    PARSE_CHECK(chunks.length < 4, "wrong number of chunks");
    CHECK_NEXT();

    final tokenId = itObj;
    PARSE_CHECK(!checkValidTokenId(tokenId ) , "tokenId invalid size");
    CHECK_NEXT();

    final List<BigInt> amounts = [];
    while (cit != chunks.length) {
      final amountBuf = itObj; //.reversed;
      PARSE_CHECK(amountBuf.length != 8, "amount string size not 8 bytes");

      final value = bufferToBN();
      amounts.add(value);

      ++cit;
      if (cit < chunks.length) {
        itObj = chunks[cit];
      }
      it = 0;
    }

    PARSE_CHECK(amounts.length == 0, "token_amounts size is 0");
    PARSE_CHECK(amounts.length > 19, "token_amounts size is greater than 19");

    final actionData = new SendParseResult(
      tokenIdBuf: tokenId,
      amounts: amounts
    );
    return new ParseResult(
      tokenType: tokenType,
      transactionType: transactionType,
      data: actionData
    );
  } else {
      PARSE_CHECK(true , "unknown action type");
  }

  // unreachable code
  return new ParseResult(
    tokenType: tokenType,
    transactionType: transactionType,
    data: new ImpossibleParseResult()
  );
}
